docs/proposals/templates.md

```diff
@@ -0,0 +1,145 @@
+  between dependent pieces
+* Support maintaining a library of templates within Kubernetes that can be accessed and instantiated by end users
+* Allow users to quickly and repeatedly deploy instances of well-defined application patterns produced by the community
+* Follow established Kubernetes API patterns by defining new template related APIs which consume+return first class Kubernetes 
+  API (and therefore json conformant) objects.
+
+We do not wish to invent a new turing complete templating language.  There are good options available 
+(e.g. https://github.com/mustache/mustache) for developers who want a completely flexible and powerful solution for creating 
+arbitrarily complex templates with parameters, and tooling can be built around such schemes.
+
+This desire for simplicity also intentionally excludes template composability/embedding as a supported use case.  
+
+Allowing templates to reference other templates presents versioning+consistency challenges along with making the template 
+no longer a self-contained portable object.  Scenarios necessitating multiple templates can be handled in one of several 
+alternate ways:
+
+* Explicitly constructing a new template that merges the existing templates (tooling can easily be constructed to perform this 
+  operation since the templates are first class api objects).
+* Manually instantiating each template and utilizing [service linking](https://github.com/kubernetes/kubernetes/pull/17543) to share 
+  any necessary configuration data.
+
+This document will also refrain from proposing server APIs or client implementations.  This has been a point of debate, and it makes
+more sense to focus on the template/parameter specification/syntax than to worry about the tooling that will process or manage the
+template objects.  However since there is a desire to at least be able to support a server side implementation, this proposal
+does assume the specification will be k8s API friendly.
+
+## Desired characteristics
+
+* Fully k8s object json-compliant syntax.  This allows server side apis that align with existing k8s apis to be constructed 
+  which consume templates and existing k8s tooling to work with them.  It also allows for api versioning/migration to be managed by
+  the existing k8s codec scheme rather than having to define/introduce a new syntax evolution mechanism.
+  * (Even if they are not part of the k8s core, it would still be good if a server side template processing+managing api supplied 
+    as an ApiGroup consumed the same k8s object schema as the peer k8s apis rather than introducing a new one)
+* Self-contained parameter definitions.  This allows a template to be a portable object which includes metadata that describe
+  the inputs it expects, making it easy to wrapper a user interface around the parameterization flow.
+* Support for non-string parameters.  K8S objects define fields of type string, int, bool, and byte[](base64).  The parameter
```

Comment:
> The only case I see is across list indices - anything else should be safe. We can use the "fake" keys for that, I think (same as patch)

i'm not clear if your paths include the actual name of the objects, or if you're just going to reference a particular replication controller within the template via ReplicationController[2], but the former case has obvious problems if someone renames the replication controller and the latter has problems if the objects in the template get reordered for any reason (eg running the file through a json formatter).  Not familiar w/ what patch does or what is meant by fake keys as a way to address this.

But aside from that, it still seems to put more burden on the template author to determine/understand the correct path reference to define, rather than simply putting "$(FOO)" exactly where they want FOO to appear.

>  For an int value - what do you store in the struct pre-expansion? How do you know the field wants to be filled in from a variable as opposed to being a litera; value?

we'd still use a "$(VALUE)" value in the field for ints, this is the reason the template schema is oblivious to the actual type of the objects within the schema, because they (the objects within the template) are not valid api objects until they get expanded (but they are valid json).

> The variable can hold a valid value (0) but you still know it needs to be patched.

ok so that is definitely valuable if we conclude that the objects within the template's object array should be valid api objects, rather than just valid json.  So far that hasn't been a requirement for us, though it certainly has benefits from a tooling and validation perspective.